components/director/pkg/operation/update_operation_handler.go

```diff
@@ -0,0 +1,154 @@
+type ResourceUpdaterFunc func(ctx context.Context, id string, ready bool, errorMsg *string) error
+
+// ResourceFetcherFunc defines a function which deletes a particular resource by ID
+type ResourceDeleterFunc func(ctx context.Context, id string) error
+
+type updateOperationHandler struct {
+	transact             persistence.Transactioner
+	resourceUpdaterFuncs map[resource.Type]ResourceUpdaterFunc
+	resourceDeleterFuncs map[resource.Type]ResourceDeleterFunc
+}
+
+// NewUpdateOperationHandler creates a new handler struct to update resource by operation
+func NewUpdateOperationHandler(transact persistence.Transactioner, resourceUpdaterFuncs map[resource.Type]ResourceUpdaterFunc, resourceDeleterFuncs map[resource.Type]ResourceDeleterFunc) *updateOperationHandler {
+	return &updateOperationHandler{
+		transact:             transact,
+		resourceUpdaterFuncs: resourceUpdaterFuncs,
+		resourceDeleterFuncs: resourceDeleterFuncs,
+	}
+}
+
+// ServeHTTP handles the Operations API requests
+func (h *updateOperationHandler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	ctx := request.Context()
+
+	if request.Method != http.MethodPost {
+		apperrors.WriteAppError(ctx, writer, apperrors.NewInternalError("Method not allowed"), http.StatusMethodNotAllowed)
+		return
+	}
+
+	bytes, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		apperrors.WriteAppError(ctx, writer, apperrors.NewInternalError("Unable to read request body"), http.StatusInternalServerError)
+		return
+	}
+	defer func() {
+		err := request.Body.Close()
+		if err != nil {
+			log.C(ctx).WithError(err).Error("Failed to close request body")
+		}
+	}()
+
+	var operation OperationRequest
+	if err := json.Unmarshal(bytes, &operation); err != nil {
+		apperrors.WriteAppError(ctx, writer, apperrors.NewInternalError("Unable to decode body to JSON"), http.StatusBadRequest)
+		return
+	}
+
+	if err := validation.ValidateStruct(&operation,
+		validation.Field(&operation.ResourceID, is.UUID),
+		validation.Field(&operation.OperationType, validation.Required, validation.In(graphql.OperationTypeCreate, graphql.OperationTypeUpdate, graphql.OperationTypeDelete)),
+		validation.Field(&operation.ResourceType, validation.Required, validation.In(resource.Application))); err != nil {
+		apperrors.WriteAppError(ctx, writer, apperrors.NewInvalidDataError("Invalid operation properties: %s", err), http.StatusBadRequest)
+		return
+	}
+
+	tx, err := h.transact.Begin()
+	if err != nil {
+		log.C(ctx).WithError(err).Errorf("An error occurred while opening db transaction: %s", err.Error())
+		apperrors.WriteAppError(ctx, writer, apperrors.NewInternalError("Unable to establish connection with database"), http.StatusInternalServerError)
+		return
+	}
+	defer h.transact.RollbackUnlessCommitted(ctx, tx)
+
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	resourceUpdaterFunc := h.resourceUpdaterFuncs[operation.ResourceType]
+	switch operation.OperationType {
```

Comment:
I see that this API was build with the undestanding that it will be invoked in the scenarios when there's nothing more to be done regarding an operation and the `application.ready` property should be flipped to `true` (whether there's an error or not).
After some discussions around the operations controller reconciliation loop I think we'll have to do a little bit more "intermediate" invocations of this API from the controller side - essentially every time the reconciliation loop ends we'll be calling this API. This is so that we can periodically reconcile the state in Director DB (the current state) with what's specified by the Operation CR (desired state). This also means that we shouldn't be flipping the `ready` property as we currently do here. I suggest extending the `OperationRequest` struct with `Ready` bool field so that the operations controller can itself dictate and specify what's the actually value of this property at a particular point in time.